### ISSUE TITLE
Failing to send traces when using flush by payload size and partial flushes

### ISSUE THREAD

**Issue by brettlangdon** [DataDog/dd-trace-py#898]
Issue copied from https://github.com/DataDog/dd-trace-py/pull/801#issuecomment-486066770

@brettlangdon I'm struggling to send traces for Django management commands, taking ~30s to ~10m to execute. I thought I'd give this branch a shot to see if it solves my problems. Unfortunately it did not, and I noticed a possibly undesirable side-effect.

Here's my log from executing the management command code interactively in a Python shell, with IP addresses and hostnames obfuscated:

```
In [2]: create_emu_data.run_create_emu_data()
2019-04-24 13:59:00,134 DEBUG writer.write:36 Queuing trace with 500 spans, 195 additional messages skipped
2019-04-24 13:59:00,814 ERROR writer._flush:201 cannot send spans to OBFUSCATED_DOMAIN:8126: [Errno 32] Broken pipe
2019-04-24 13:59:00,823 DEBUG writer._target:155 Popped trace with 500 spans, 43 additional messages skipped
2019-04-24 13:59:00,835 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107507b, full=False), 6 additional messages skipped
2019-04-24 13:59:00,838 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107507b, full=False), 6 additional messages skipped
2019-04-24 13:59:00,843 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107508b, 6 additional messages skipped
2019-04-24 13:59:01,992 DEBUG api._put:196 Response 200 OK length:488, 5 additional messages skipped
2019-04-24 13:59:01,997 DEBUG api.send_traces:172 Reported 1 traces in 1.15439s, 5 additional messages skipped
2019-04-24 13:59:04,285 DEBUG span.set_tag:138 error setting numeric metric _dd1.sr.eausr:None, 18 additional messages skipped

In [3]: 2019-04-24 13:59:57,735 DEBUG sampler.__init__:41 initialized RateSampler, sample 100% of traces, 6 additional messages skipped
2019-04-24 14:00:00,454 DEBUG api._put:196 Response 200 OK length:521, 43 additional messages skipped
2019-04-24 14:00:00,454 DEBUG api.send_traces:172 Reported 1 traces in 1.37156s, 43 additional messages skipped
2019-04-24 14:00:00,455 DEBUG writer._target:155 Popped trace with 500 spans, 44 additional messages skipped
2019-04-24 14:00:00,457 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107503b, full=False), 44 additional messages skipped
2019-04-24 14:00:00,457 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107503b, full=False), 44 additional messages skipped
2019-04-24 14:00:00,457 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107504b, 44 additional messages skipped
2019-04-24 14:00:05,100 DEBUG sampler.__init__:41 initialized RateSampler, sample 100% of traces
In [3]:

In [3]: 2019-04-24 14:01:00,554 DEBUG api._put:196 Response 200 OK length:674, 40 additional messages skipped
2019-04-24 14:01:00,554 DEBUG api.send_traces:172 Reported 1 traces in 2.08556s, 40 additional messages skipped
2019-04-24 14:01:00,556 DEBUG writer._target:155 Popped trace with 500 spans, 40 additional messages skipped
2019-04-24 14:01:00,563 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107503b, full=False), 40 additional messages skipped
2019-04-24 14:01:00,564 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107503b, full=False), 40 additional messages skipped
2019-04-24 14:01:00,564 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107504b, 40 additional messages skipped
2019-04-24 14:01:28,364 DEBUG api.get_json:86 Unable to parse Datadog Agent JSON response: Expecting value: line 1 column 1 (char 0) 'read tcp OBFUSCATED_IP:8126->OBFUSCATED_IP:62946: i/o timeout\n'
2019-04-24 14:01:28,365 ERROR writer._log_error_status:218 failed_to_send traces to Datadog Agent: HTTP error status 400, reason Bad Request, message Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Wed, 24 Apr 2019 04:31:28 GMT
Content-Length: 59


In [3]:
Do you really want to exit ([y]/n)?
2019-04-24 14:01:58,199 DEBUG span.set_tag:138 error setting numeric metric _dd1.sr.eausr:None, 103 additional messages skipped
2019-04-24 14:01:58,200 DEBUG writer.write:36 Queuing trace with 1 spans, 194 additional messages skipped
2019-04-24 14:01:58,410 DEBUG writer._on_shutdown:115 Waiting 5s for traces to be sent. Hit ctrl-c to quit.
2019-04-24 14:02:01,175 DEBUG api._put:196 Response 200 OK length:673, 32 additional messages skipped
2019-04-24 14:02:01,176 DEBUG api.send_traces:172 Reported 1 traces in 1.73121s, 32 additional messages skipped
2019-04-24 14:02:01,177 DEBUG writer._target:155 Popped trace with 500 spans, 32 additional messages skipped
2019-04-24 14:02:01,181 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107503b, full=False), 32 additional messages skipped
2019-04-24 14:02:01,181 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107503b, full=False), 32 additional messages skipped
2019-04-24 14:02:01,181 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107504b, 32 additional messages skipped
2019-04-24 14:03:00,046 DEBUG api._put:196 Response 200 OK length:675, 37 additional messages skipped
2019-04-24 14:03:00,047 DEBUG api.send_traces:172 Reported 1 traces in 1.36665s, 37 additional messages skipped
2019-04-24 14:03:00,048 DEBUG writer._target:155 Popped trace with 500 spans, 37 additional messages skipped
2019-04-24 14:03:00,051 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107505b, full=False), 37 additional messages skipped
2019-04-24 14:03:00,052 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107505b, full=False), 37 additional messages skipped
2019-04-24 14:03:00,052 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107506b, 37 additional messages skipped
2019-04-24 14:04:02,631 DEBUG api._put:196 Response 200 OK length:554, 42 additional messages skipped
2019-04-24 14:04:02,632 DEBUG api.send_traces:172 Reported 1 traces in 3.70227s, 42 additional messages skipped
2019-04-24 14:04:02,633 DEBUG sampler.__init__:41 initialized RateSampler, sample 25.329419826708303% of traces, 2 additional messages skipped
2019-04-24 14:04:02,633 DEBUG writer._target:155 Popped trace with 500 spans, 42 additional messages skipped
2019-04-24 14:04:02,634 DEBUG writer._target:162 Attempting to flush Payload(length=1, size=107505b, full=False), 42 additional messages skipped
2019-04-24 14:04:02,635 DEBUG writer._flush:187 Flushing payload Payload(length=1, size=107505b, full=False), 42 additional messages skipped
2019-04-24 14:04:02,635 DEBUG api._put:187 PUT OBFUSCATED_DOMAIN:8126/v0.4/traces with payload 107506b, 42 additional messages skipped
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/OBFUSCATED/src/ddtrace/ddtrace/writer.py", line 118, in _on_shutdown
    self._trace_queue.join()
  File "/Users/OBFUSCATED/src/ddtrace/ddtrace/writer.py", line 334, in join
    self.not_full.wait(0.05)
  File "/Users/OBFUSCATED/.pyenv/versions/3.6.6/lib/python3.6/threading.py", line 299, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt

$ 
```

I was already getting the broken pipe errors before I tried this branch. I had hoped this branch would fix them. And other traces are being sent correctly (`django.request`, `sqlite`, etc.) So I thought it was just related to the size of the trace for this long running code.

But the possibly undesirable thing I noticed with this branch (or at least a different thing vs ddtrace-py 0.24) is that it keeps trying to send the spans again and again after the command has finished executing, and when I attempt to exit the interactive shell it says it will try for 5s, but it actually just keeps trying forever until `KeyboardInterrupt`.

I also set `export DD_TRACER_PARTIAL_FLUSH_ENABLED=true` before opening this Python shell, which didn't seem to make any difference.

**Comment by brettlangdon** [comment]
@mrmachine can you describe what the trace is doing exactly? I assume it is a cli script that is making a ton of database calls?

One thing worth noting that we have found is that traces that produce a ton of spans (say 2k or more spans) tend to be extremely difficult to read, it might be worth evaluating the structure of this trace to determine if there is any way to simplify it.

The background worker not closing after 5 seconds is a good find, I am going to dig into this further.

Just to help define some terminology:

**Flush by payload size:**

When a trace is closed (all spans have finished) the tracer will push the trace into a queue for a background worker to pick up and flush to our agent.

Currently (without #801) we buffer all of those traces in memory and then flush them once every second. The problem here is if the system is under high load we might buffer more traces than a single API call to the agent will allow (10mb) which causes the request to be rejected.

With #801 we want to move to a new strategy of flushing by the payload size, so instead when we buffer the traces into memory in the background worker we keep track of the encoded size of that trace, once we reach enough traces that a size threshold is reached we will flush the traces to the agent. If the size threshold isn't reached then we will continue to flush every 1 second like before. So the new semantic is "when size limit is reached or once every second, whichever comes first".

**Partial flushes:**

This feature allows a single trace to be broken into smaller traces. This is typically useful for large or long running processes that generate a ton of spans. The idea here is that as soon as a trace has > 500 closed spans, we will separate those closed spans into their own trace and push them into the background worker to be flushed, keeping the original trace/opened spans still alive and generating spans.

**Comment by mrmachine** [comment]
@brettlangdon Yes, it is a CLI script that is making a tonne of DB calls (and doing a tonne of data transformation in Python).

The script is executed very infrequently, but it's very slow and I am trying to optimise it. I don't really need to collect a tonne of traces for this script. Just one at a time, so I can identify N+1 queries in a loop and also bottlenecks in Python in between DB calls. Then I'd make some changes, re-run the script, and look at the new trace again.

I even tried breaking it down into smaller traces by removing `@tracer.wrap()` from the main function, but keeping it on the other functions that are called by the main function, but those traces still did not show up in Datadog.

Plus, I think it will be harder to mentally piece these individual traces back together to get a high level overview. To properly break it down, I think I'd need to see the complete flame graph for at least a few iterations to identify the major components and then I could start sending smaller traces for the problematic areas of code only. But by then, I've probably already optimised the script and no longer need to collect much trace data at all, anyway.

I think I understand the terminology you have described. But I don't understand why I'm still getting broken pipe errors (same on 0.24 and this branch). I did also run this code on a staging server as well where it executed faster (20s instead of 30s), and I got a 413 payload too large error that time. Could the broken pipe error be related to greater latency or reduced bandwidth between my development machine (in AU) and the server where the APM enabled datadog agent is running (in the US), vs the staging server (also located in AU, but on AWS and not a home connection). My home connection is fibre with a consistent 90mbps down / 30mbps up, though.

**Comment by brettlangdon** [comment]
@mrmachine all makes sense.

Let me think on this a bit and try to see if I can get a similar setup locally to figure out how best to address.

I am not 100% sure about the broken pipe issue, it could be due to latency, most of the time people run the datadog agent locally so our connection timeouts are usually kept pretty low.

**Comment by mrmachine** [comment]
@brettlangdon I just reviewed sentry errors on our production environment where Datadog agent is running on the same host as ddtrace-py. In 24h we have 78x `cannot send spans to host.docker.internal:8126: [Errno 32] Broken pipe` errors and 214x `failed_to_send traces to Datadog Agent: HTTP error status 413...` errors. But this is using `ddtrace==0.23.0` and without `export DD_TRACER_PARTIAL_FLUSH_ENABLED=true`.

I believe I could probably resolve the 413 errors with `export DD_TRACER_PARTIAL_FLUSH_ENABLED=true` (but this would make some stats report higher than actual req/s)?

But the broken pipe issue is the same issue I reported with ddtrace-py running on a different host and you mentioned this might not be a problem with the usual setup where the agent runs on the same host as the client.

**Comment by brettlangdon** [comment]
@mrmachine for the issue for the payloads being too large (413 errors) should be resolved by #941 and released in version `0.27.0`, have you had a chance to try that version out for that specific error?

I assume regardless of the 413 errors you are still having issues with broken pipes?

**Comment by majorgreys** [comment]
Closing due to inactivity. @mrmachine if this is still an issue, please reopen.
